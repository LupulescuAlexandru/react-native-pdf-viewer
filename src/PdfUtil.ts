import { NativeModules } from 'react-native';

export type PageDim = { height: number; width: number };

type PdfUtilType = {
  /**
   * Get the number of pages of a pdf.
   */
  getPageCount(source: string): Promise<number>;

  /**
   * Get the dimensions of every page.
   */
  getPageSizes(source: string): Promise<PageDim[]>;
};

/**
 * Utility pdf actions.
 */
export const PdfUtil: PdfUtilType = NativeModules.RNPdfUtil;
