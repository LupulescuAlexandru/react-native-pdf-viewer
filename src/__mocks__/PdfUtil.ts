export const PdfUtil = {
  getPageCount: jest.fn(),
  getPageSizes: jest.fn(),
};
