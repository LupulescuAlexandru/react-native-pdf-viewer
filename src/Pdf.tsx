import React, { useEffect, useState } from 'react';
import { FlatList, StyleSheet, View } from 'react-native';

import { PageDim, PdfUtil } from './PdfUtil';
import { PdfView } from './PdfView';

type PdfProps = {
  /**
   * Callback to handle errors.
   */
  onError?: (error: Error) => void;

  /**
   * Callback to handle pdf load completion.
   *
   * Passed the page count of the loaded pdf.
   */
  onLoadComplete?: (numberOfPages: number) => void;

  /**
   * Document to display.
   */
  source: string;
};

const separatorSize = 4;

/**
 * Display a pdf.
 */
export function Pdf({ onError, onLoadComplete, source }: PdfProps) {
  const [flatListLayout, setFlatListLayout] = useState<PageDim>({
    height: 0,
    width: 0,
  });
  const [pageDims, setPageDims] = useState<PageDim[]>([]);
  useEffect(() => {
    const state = { live: true };
    PdfUtil.getPageSizes(source)
      .then((sizes) => {
        if (state.live) {
          setPageDims(sizes);
          if (onLoadComplete != null) {
            onLoadComplete(sizes.length);
          }
        }
      })
      .catch((error) => {
        if (state.live && onError != null) {
          onError(error);
        }
      });

    return () => {
      state.live = false;
    };
  }, [onError, onLoadComplete, setPageDims, source]);

  return (
    <FlatList
      data={pageDims}
      getItemLayout={(data, index) => {
        let itemHeight = 100;
        let offset = (itemHeight + separatorSize) * index;
        if (data == null) {
          console.warn('Pdf list getItemLayout() not passed data.');
        } else if (flatListLayout.height === 0 || flatListLayout.width === 0) {
          console.warn(
            'Pdf list getItemLayout() could not determine screen size.'
          );
        } else {
          let pageSize = data[index];
          itemHeight =
            (flatListLayout.width * pageSize.height) / pageSize.width;
          offset = 0;
          for (let i = 0; i < index; ++i) {
            pageSize = data[i];
            offset +=
              separatorSize +
              (flatListLayout.width * pageSize.height) / pageSize.width;
          }
        }
        return {
          length: itemHeight,
          offset,
          index,
        };
      }}
      initialNumToRender={1}
      ItemSeparatorComponent={() => <View style={styles.separator} />}
      keyExtractor={(_item, index) => index.toString()}
      maxToRenderPerBatch={2}
      onLayout={(event) => {
        setFlatListLayout({
          height: event.nativeEvent.layout.height,
          width: event.nativeEvent.layout.width,
        });
      }}
      renderItem={({ index }) => (
        <PdfView page={index} source={source} style={styles.page} />
      )}
      windowSize={5}
    />
  );
}

const styles = StyleSheet.create({
  page: {
    backgroundColor: '#fff',
    elevation: 4,
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.23,
    shadowRadius: 2.62,
  },
  separator: { margin: separatorSize },
});
