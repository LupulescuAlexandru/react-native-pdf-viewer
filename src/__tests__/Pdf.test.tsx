import React from 'react';
import { act, render } from '@testing-library/react-native';
import { Pdf } from '../Pdf';
import { PdfUtil } from '../PdfUtil';

jest.mock('../PdfUtil');
const MockPdfUtil = PdfUtil as jest.Mocked<typeof PdfUtil>;

function promisify(cb: jest.Mock) {
  return new Promise((resolve) => cb.mockImplementationOnce(resolve));
}

test('renders pdf viewer', async () => {
  MockPdfUtil.getPageSizes.mockImplementationOnce(() =>
    Promise.resolve([
      { height: 1, width: 1 },
      { height: 1, width: 1 },
    ])
  );
  const onLoadComplete = jest.fn();
  const loadCompletePromise = promisify(onLoadComplete);

  const { toJSON } = render(
    <Pdf onLoadComplete={onLoadComplete} source="test.pdf" />
  );

  expect(toJSON()).toMatchSnapshot();

  await act(async () => await loadCompletePromise);
  expect(onLoadComplete).toHaveBeenCalledWith(2);
});

test('pdf viewer load error', async () => {
  MockPdfUtil.getPageSizes.mockImplementationOnce(() => Promise.reject());
  const onError = jest.fn();
  const errorPromise = promisify(onError);
  const onLoadComplete = jest.fn();
  const loadCompletePromise = promisify(onLoadComplete);

  render(
    <Pdf onError={onError} onLoadComplete={onLoadComplete} source="test.pdf" />
  );

  await act(
    async () => await Promise.race([errorPromise, loadCompletePromise])
  );
  expect(onError).toHaveBeenCalled();
  expect(onLoadComplete).not.toHaveBeenCalled();
});

async function getRenderedViewer() {
  MockPdfUtil.getPageSizes.mockImplementationOnce(() =>
    Promise.resolve([
      { height: 400, width: 300 },
      { height: 400, width: 300 },
    ])
  );
  const onLoadComplete = jest.fn();
  const loadCompletePromise = promisify(onLoadComplete);
  const result = render(
    <Pdf onLoadComplete={onLoadComplete} source="test.pdf" />
  );
  await act(async () => await loadCompletePromise);
  return result;
}

test('viewer renders pages', async () => {
  const { getByTestId } = await getRenderedViewer();
  const flatList = getByTestId('pdfFlatList');

  const { toJSON } = render(flatList.props.renderItem({ index: 0 }));

  expect(toJSON()).toMatchSnapshot();
});

test('viewer computes layout', async () => {
  const { getByTestId } = await getRenderedViewer();
  const flatList = getByTestId('pdfFlatList');
  const pageDimsData = [
    { height: 400, width: 300 },
    { height: 400, width: 300 },
  ];

  expect(flatList.props.data).toEqual([]);

  act(() =>
    flatList.props.onLayout({
      nativeEvent: { layout: { height: 1000, width: 600 } },
    })
  );

  expect(flatList.props.data).toEqual(pageDimsData);
  expect(flatList.props.getItemLayout(pageDimsData, 1)).toEqual({
    index: 1,
    length: 800,
    offset: 808,
  });
});

test('viewer measures pages', async () => {
  MockPdfUtil.getPageSizes.mockImplementationOnce(() =>
    Promise.resolve([
      { height: 400, width: 300 },
      { height: 400, width: 300 },
      { height: 300, width: 150 },
    ])
  );
  const onMeasurePages = jest.fn();
  const measurePagesPromise = promisify(onMeasurePages);

  const { getByTestId } = render(
    <Pdf onMeasurePages={onMeasurePages} source="test.pdf" />
  );
  const flatList = getByTestId('pdfFlatList');

  act(() =>
    flatList.props.onLayout({
      nativeEvent: { layout: { height: 1000, width: 600 } },
    })
  );

  await act(async () => await measurePagesPromise);
  expect(onMeasurePages).toHaveBeenCalledWith([
    { itemHeight: 800, offset: 0 },
    { itemHeight: 800, offset: 808 },
    { itemHeight: 1200, offset: 1616 },
  ]);
});
