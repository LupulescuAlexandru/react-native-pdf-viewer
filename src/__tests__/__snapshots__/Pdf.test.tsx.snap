// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders pdf viewer 1`] = `
<RCTScrollView
  ItemSeparatorComponent={[Function]}
  data={Array []}
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  getItemLayout={[Function]}
  horizontal={false}
  initialNumToRender={1}
  keyExtractor={[Function]}
  maxToRenderPerBatch={2}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  testID="pdfFlatList"
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={5}
>
  <View />
</RCTScrollView>
`;

exports[`viewer renders pages 1`] = `
<RNPdfView
  onPdfError={[Function]}
  onPdfLoadComplete={[Function]}
  page={0}
  source="test.pdf"
  style={
    Object {
      "backgroundColor": "#fff",
      "elevation": 4,
      "shadowColor": "#000",
      "shadowOffset": Object {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 0.23,
      "shadowRadius": 2.62,
    }
  }
/>
`;
